feat: swiftformat.path workspace-local overrides

maximkrouk opened the pull request:
I'm not sure if it works, so it would be cool if you could check it, but the idea is to resolve relative paths so that users of this extension will be able to set project-specific executables.

Current implementation should try to resolve local path and in case it fails (no matching local path found) it will run fallback logic for the global path.

maximkrouk committed:
feat: swiftformat.path local override support

vknabel commented:
While I get the use case for this PR, I am not really convinced regarding the potential security risk vs actual use tradeoff.

In my opinion project specifics should not require global solutions that might conflict with other projects.

Instead you could create a global shell script that performs the check up and decides which path of swiftformat to use and set `swiftformat.path` accordingly to that script.

If https://github.com/vknabel/vscode-swiftformat/issues/59#issuecomment-2934932260 is correct, could you have a look what it does different there? In theory the 3 extensions for swiftformat, apple's swift-format and SwiftLint should behave similarly. If there is such a central difference, this is likely a bug or unwanted inconsistency.

In the end take all of this with a grain of salt as I am no longer part of the swift or vscode communities and don't use any of these anymore. Also I will only be notified by pull requests, not issues.

I can review this, but cannot provide any guarantees when or if I do as it requires me quite a lot of ceremony to set everything up before the review and frankly does not provide any benefit for me.

I hope you understand and that I could be a little bit helpful here and there.

SimplyDanny commented:
The decisive difference to the SwiftLint extension is [this line](https://github.com/vknabel/vscode-swiftformat/blob/67e86b7dbf69ed6cecccdaf6592baa572e94f905/package.json#L73). Changing it to `machine-overridable` here will make the option available to project/workspace scopes as well. 

At the moment, local paths work, but they need to be configured centrally in the user `settings.json` which is a bit odd.

SimplyDanny reviewed (changes requested):
Would you like to the make the necessary adaptions in the package manifest, @maximkrouk?

The proposed changes are not required to make it work.

maximkrouk commented:
I guess if changes in package manifest are sufficient (those changes are at least necessary) and path option allows to set relative path to swiftformat binary than it's much better to just introduce mentioned manifest mods.

From readme and high level code overview it looked like only global binaries are supported 🤔

SimplyDanny commented:
If you set an explicit (local) path, [it will be preferred](https://github.com/vknabel/vscode-swiftformat/blob/67e86b7dbf69ed6cecccdaf6592baa572e94f905/src/Current.ts#L127).

maximkrouk commented:
> _If you set an explicit (local) path, [it will be preferred](https://github.com/vknabel/vscode-swiftformat/blob/67e86b7dbf69ed6cecccdaf6592baa572e94f905/src/Current.ts#L127)._

Yes, but afaiu it returns string as it is and is used [here](https://github.com/vknabel/vscode-swiftformat/blob/67e86b7dbf69ed6cecccdaf6592baa572e94f905/src/SwiftFormatEditProvider.ts#L46) so the path should be global, not workspace-related 🤔

I'm a bit confused because from looking at code it shouldn't work with local paths, but even release version seem to work (with global settings override) with workspace-local path 🌚

maximkrouk commented:
I reverted changes in `Current.ts`, modified `package.json` and updated readme so it is a bit clearer that users can specify workspace-related path
But I'm still not sure how to test it locally

SimplyDanny commented:
> > _If you set an explicit (local) path, [it will be preferred](https://github.com/vknabel/vscode-swiftformat/blob/67e86b7dbf69ed6cecccdaf6592baa572e94f905/src/Current.ts#L127)._
> 
> Yes, but afaiu it returns string as it is and is used [here](https://github.com/vknabel/vscode-swiftformat/blob/67e86b7dbf69ed6cecccdaf6592baa572e94f905/src/SwiftFormatEditProvider.ts#L46) so the path should be global, not workspace-related 🤔

The code you are referring to falls back to `swiftformat.path` eventually in its implementation. The returned local/relative path is finally called and is not resolved globally. At least for me, it looks reasonable at the moment.

maximkrouk commented:
I guess it should be fine now to be merged and released 👉👈

vknabel commented:
Great work from you two!

@SimplyDanny I trust you, if it's fine for you, it's fine for me. Feel free to merge and release PRs. :)
